### PR TITLE
Automation of releases

### PR DESCRIPTION
**Pull request by maxschulz-COL**
## Description

**Automation of releases**
- only requires `hatch run prep-release` to prepare a release
- then PR needs to be raised and approved
- on merge, we tag main, publish on PyPI, create a github release and create a PR that bumps the version
- the version bump still needs to be approved manually

**Potential future improvements**
- auto-merge version bump, or even better, use `hatch-vcs`

## Checklist

- [x] I have not referenced individuals, products or companies in any commits, directly or indirectly
- [x] I have not added data or restricted code in any commits, directly or indirectly
- [x] I have updated the docstring of any public function/class/model changed
- [ ] I have added the PR number to the change description in the changelog fragment, e.g. `Enable feature XXX (#123)` (if applicable)

## Types of changes

- [x] Docs/refactoring (non-breaking change which improves codebase)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: tools/check_package_release.py
Comment (petar-qb) [inline tools/check_package_release.py:58]: Do we need to check `if double_release` **before** we overwrite the `env_file`:
```
if double_release:
    sys.exit("Cannot release two packages at the same time. Please modify your PR.")

with open(env_file, "a") as f:
    f.write(f"NEW_RELEASE={new_release}\n")
    ...
```

File: tools/check_package_release.py
Comment (petar-qb) [inline tools/check_package_release.py:32]: Why are we processing the `new_release` value with a string instead of a bool.

File: .github/workflows/check-release.yml
Comment (petar-qb) [inline .github/workflows/check-release.yml:27]: You've prepared this configuration so it can handle a process with multiple packages. Do we already have the potential candidate that's not `vizro-core` and did you test it with multiple packages?

File: .github/workflows/check-release.yml
Comment (petar-qb) [inline .github/workflows/check-release.yml:107]: And what if we want to release a `minor` version rather than a `patch` version?

File: .github/workflows/check-release.yml
Comment (petar-qb) [inline .github/workflows/check-release.yml:111]: Are we okay with having the "maxschulz-COL" in the email here?

**Review by petar-qb — COMMENTED** [review]
Awesome work 🚀 🚀!

File: vizro-core/docs/pages/development/contributing.md
Comment (l0uden) [inline vizro-core/docs/pages/development/contributing.md:152]: Does it mean that `prep-release` will create this version? I've created branch `release/v1.0.0`, but it created `0.1.1`

**Review by l0uden — CHANGES_REQUESTED** [review]

File: tools/check_package_release.py
Comment (maxschulz-COL) [reply tools/check_package_release.py:58]: I think it doesn't matter, but it may be clearer the other way round

File: tools/check_package_release.py
Comment (maxschulz-COL) [reply tools/check_package_release.py:32]: good point - it was taken over from a script in kedro, let me try to change it

File: .github/workflows/check-release.yml
Comment (maxschulz-COL) [reply .github/workflows/check-release.yml:27]: Yes, the work of AX will be the next package, and I tested it with a hypothetical next package, but I expect that there will still be slight adjustments necessary when it comes to that

File: .github/workflows/check-release.yml
Comment (maxschulz-COL) [reply .github/workflows/check-release.yml:107]: Then we have to change the version manually, but this is likely happening only when a breaking PR hits main, so it should be done with that PR instead of right after release of the previous release

File: .github/workflows/check-release.yml
Comment (maxschulz-COL) [reply .github/workflows/check-release.yml:111]: I followed the github privacy guide, this is essentially the maximum you can do to hide your email. It is not my real email

File: .github/workflows/check-release.yml
Comment (maxschulz-COL) [reply .github/workflows/check-release.yml:12]: Do I also update the `setup-python@v3` to `v4` or just the above?

File: .github/workflows/check-release.yml
Comment (maxschulz-COL) [reply .github/workflows/check-release.yml:110]: But the lint workflow does not modify the code right?

File: vizro-core/docs/pages/development/contributing.md
Comment (maxschulz-COL) [reply vizro-core/docs/pages/development/contributing.md:152]: No, the name is arbitrary, let me adjust the docs then

File: .github/workflows/check-release.yml
Comment (l0uden) [reply .github/workflows/check-release.yml:110]: It will modify the code, but what I'm thinking of now is why we even need `linting` here? This workflow will run only after merging to `main`, in this case all this checks should be green already. And also here's the logic if first lint fail then run it again and then push the changes is smth we should be avoid, because this check is present in the different workflow.

I think it is better to forget about my fist comment. What I'm thinking of is that this linting step is unnecessary here or anywhere (I saw it in `hatch.toml`) except lint workflow.

File: .github/workflows/check-release.yml
Comment (l0uden) [reply .github/workflows/check-release.yml]: Yes, it also has `v4` in other workflows. I didn't notice)

File: .github/workflows/check-release.yml
Comment (maxschulz-COL) [reply .github/workflows/check-release.yml:110]: The reason why I thought it was necessary is because we are modifying files (`hatch version` and `hatch schema`) and we want the new PR that is raised, to be merable immediately.

I tried this out now, and it seems like that it doesn't need linting, but it did for me before. Let me take it out and see

File: vizro-core/changelog.d/20230915_010825_maximilian_schulz_prepare_automation.md
Comment (lingyielia) [inline vizro-core/changelog.d/20230915_010825_maximilian_schulz_prepare_automation.md:14]: Do we want to insert content here?

File: vizro-core/docs/pages/development/contributing.md
Comment (lingyielia) [inline vizro-core/docs/pages/development/contributing.md:152]: So the branch name would be matched with what's in the `__init__.py` right? Like currently it's `"0.1.1.dev0"` so the branch name should be release/v0.1.1

Shall we include the package name into the branch name as well? something like `release/vizro-core/v0.1.1`

File: vizro-core/docs/pages/development/contributing.md
Comment (lingyielia) [inline vizro-core/docs/pages/development/contributing.md:161]: Maybe the merging queue could be useful due to this potential conflicts? https://github.blog/2023-07-12-github-merge-queue-is-generally-available/

Just an idea, definitely no need to pivot our current approach now

**Review by lingyielia — CHANGES_REQUESTED** [review]

File: vizro-core/changelog.d/20230915_010825_maximilian_schulz_prepare_automation.md
Comment (maxschulz-COL) [reply vizro-core/changelog.d/20230915_010825_maximilian_schulz_prepare_automation.md:14]: I thought about it, but I think it is not really relevant to the user right?

File: vizro-core/docs/pages/development/contributing.md
Comment (maxschulz-COL) [reply vizro-core/docs/pages/development/contributing.md:161]: Good shout, I'll take a closer look. Thanks for sharing

File: tools/check_package_release.py
Comment (Anna-Xiong) [inline tools/check_package_release.py:36]: if we do a check here similar check like:
```
    if len(sys.argv) != ARG_NUM:
        raise Exception("Usage: python extract_release_notes.py <filename> <heading>")
```

would the message be printed out if someone used it run when edit CICD process? if so, would it be helpful to have exception with usage instruction here as well

**Review by lingyielia — APPROVED** [review]
LGTM! Thanks for sorting this out 💯

**Review by petar-qb — APPROVED** [review]
Thanks for addressing and answering all the comments. Looks good!

File: vizro-core/docs/pages/development/contributing.md
Comment (maxschulz-COL) [reply vizro-core/docs/pages/development/contributing.md:150]: tbh you are right - I removed it and added it in the maintainer development docs

**Review by petar-qb — APPROVED** [review]
👍

**Review by l0uden — DISMISSED** [review]

File: tools/check_package_release.py
Comment (petar-qb) [inline tools/check_package_release.py:11]: Is it used anywhere?

**Review by petar-qb — APPROVED** [review]
One comment left. Besides, it looks good!

File: .github/workflows/check-release.yml
Comment (lingyielia) [inline .github/workflows/check-release.yml:36]: Like how the final solution laid out 👍

**Review by Anna-Xiong — APPROVED** [review]
LGTM! thanks for addressing the comments!

**Review by petar-qb — APPROVED** [review]
⚡

**Pull request merged**
